pallets/nft/src/lib.rs

```diff
@@ -0,0 +1,297 @@
+        DuplicateMetadataKey,
+        /// The associated asset is not of type NFT.
+        InvalidAssetType,
+        /// Either the number of keys or the key identifier does not match the keys defined for the collection.
+        InvalidMetadataAttribute,
+        /// The maximum number of metadata keys was exceeded.
+        MaxNumberOfKeysExceeded,
+        /// The NFT does not exist.
+        NFTNotFound,
+        /// At least one of the metadata keys has not been registered.
+        UnregisteredMetadataKey,
+    }
+}
+
+impl<T: Config> Module<T> {
+    fn base_create_nft_collection(
+        origin: T::Origin,
+        ticker: Ticker,
+        collection_keys: NFTCollectionKeys,
+    ) -> DispatchResult {
+        // Verifies if the caller has the right permissions to create the collection
+        let _did = Identity::<T>::ensure_perms(origin.clone())?;
+
+        // Verifies if the ticker is already associated to an NFT collection
+        if CollectionTicker::contains_key(&ticker) {
+            return Err(Error::<T>::CollectionAlredyRegistered.into());
+        }
+
+        // Verifies if the maximum number of keys is respected and that all keys have been registered
+        if collection_keys.len() > T::MaxNumberOfCollectionKeys::get() as usize {
+            return Err(Error::<T>::MaxNumberOfKeysExceeded.into());
+        }
+
+        // Returns an error in case a duplicated key is found
+        let n_keys = collection_keys.len();
+        let collection_keys: BTreeSet<AssetMetadataKey> = collection_keys.into_iter().collect();
+        if n_keys != collection_keys.len() {
+            return Err(Error::<T>::DuplicateMetadataKey.into());
+        }
+
+        for key in &collection_keys {
+            if !Asset::<T>::check_asset_metadata_key_exists(&ticker, key) {
+                return Err(Error::<T>::UnregisteredMetadataKey.into());
+            }
+        }
+
+        // Verifies if the asset is of type NFT or creates an nft asset if it does not exist
+        match Asset::<T>::nft_asset(&ticker) {
+            Some(is_nft_asset) => {
+                if !is_nft_asset {
+                    return Err(Error::<T>::InvalidAssetType.into());
+                }
+            }
+            None => Asset::<T>::create_asset(
+                origin,
+                AssetName(ticker.as_slice().to_vec()),
+                ticker.clone(),
+                false,
+                AssetType::NFT,
+                Vec::new(),
+                None,
+                false,
+            )?,
+        }
+
+        // Creates the nft collection
+        let collection_id = NextCollectionId::try_mutate(try_next_pre::<T, _>)?;
+        let nft_collection = NFTCollection::new(collection_id.clone(), ticker.clone());
+        Collection::insert(collection_id.clone(), nft_collection);
+        CollectionKeys::insert(collection_id.clone(), collection_keys);
+        CollectionTicker::insert(ticker.clone(), collection_id.clone());
```

Comment:
The Substrate storage methods can also take a reference which avoids the need for cloning.

For both the key and value they take an `EncodeLike<T>` which is also implemented for `&T`.   (See https://paritytech.github.io/substrate/master/frame_support/storage/trait.StorageMap.html#tymethod.insert)

```suggestion
        Collection::insert(&collection_id, nft_collection);
        CollectionKeys::insert(&collection_id, collection_keys);
        CollectionTicker::insert(&ticker, &collection_id);
```